api/proto/teleport/devicetrust/v1/devicetrust_service.proto

```diff
@@ -349,6 +358,80 @@ message MacOSEnrollChallengeResponse {
   bytes signature = 2;
 }
 
+// The payload containing TPM specific information required on device
+// enrollment.
+message TPMEnrollPayload {
+  oneof ek {
+    // The device's endorsement certificate in X509, ASN.1 DER form. This
+    // certificate contains the public key of the endorsement key. This is
+    // preferred to ek_key.
+    bytes ek_cert = 1;
+    // The device's public endorsement key in PKIX, ASN.1 DER form. This is
+    // used when a TPM does not contain any endorsement certificates.
+    bytes ek_key = 2;
+  }
+  // The attestation key and the parameters necessary to remotely verify it as
+  // related to the endorsement key.
+  TPMAttestationParameters attestation_parameters = 3;
+  // The application key and the parameters necessary to remotely certify it as
+  // related to the attestation key.
+  TPMCertificationParameters application_certification_parameters = 4;
+}
+
+// The attestation key and the parameters necessary to remotely verify it as
+// related to the endorsement key.
+// See https://pkg.go.dev/github.com/google/go-attestation/attest#AttestationParameters.
+// This message excludes the `UseTCSDActivationFormat` field from the link above
+// as it is TMP 1.x specific and always false.
+message TPMAttestationParameters {
+  // The encoded TPMT_PUBLIC structure containing the attestation public key
+  // and signing parameters.
+  bytes public = 1;
+  // The properties of the attestation key, encoded as a TPMS_CREATION_DATA
+  // structure.
+  bytes create_data = 2;
+  // An assertion as to the details of the key, encoded as a TPMS_ATTEST
+  // structure.
+  bytes create_attestation = 3;
+  // A signature of create_attestation, encoded as a TPMT_SIGNATURE structure.
+  bytes create_signature = 4;
+}
+
+// The application key and the parameters necessary to remotely certify it as
+// related to the attestation key.
+// See https://pkg.go.dev/github.com/google/go-attestation/attest#CertificationParameters
+message TPMCertificationParameters {
+  // The encoded TPMT_PUBLIC structure containing the application public key
+  // and signing parameters.
+  bytes public = 1;
+  // The properties of the application key, encoded as a TPMS_CREATION_DATA
+  // structure.
+  bytes create_data = 2;
+  // An assertion as to the details of the key, encoded as a TPMS_ATTEST
+  // structure.
+  bytes create_attestation = 3;
+  // A signature of create_attestation, encoded as a TPMT_SIGNATURE structure.
+  bytes create_signature = 4;
+}
+
+// The enrollment challenge sent to the client by the server in order to
+// validate it has possession of the EK and AK.
+message TPMEnrollChallenge {
+  // A handle to the attestation key that should be used for the challenge.
+  bytes credential = 1;
```

Comment:
Remind me: does the handle have to come from the server? Wouldn't it match information the client already holds?

I know I spec-ed this myself some time ago, but my memory is failing me now.

Reply:
7c70ced9cb should provide some additional clarity to this.